File: plugins/OrchidPresentations/src/main/java/com/eden/orchid/presentations/PresentationsGenerator.java
package com.eden.orchid.presentations;

import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.generators.OrchidGenerator;
import com.eden.orchid.api.options.Option;
import com.eden.orchid.api.options.annotations.StringDefault;
import com.eden.orchid.api.resources.resource.OrchidResource;
import com.eden.orchid.api.theme.pages.OrchidPage;
import com.eden.orchid.utilities.OrchidUtils;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PresentationsGenerator extends OrchidGenerator {

    private final PresentationsModel model;

    @Option("baseDir")
    @StringDefault("presentations")
    public String presentationsBaseDir;

    @Option("presentations")
    public String[] sectionNames;

    @Inject
    public PresentationsGenerator(OrchidContext context, PresentationsModel model) {
        super(context, "presentations", 25);
        this.model = model;
    }

    @Override
    public String getDescription() {
        return "";
    }

    @Override
    public List<? extends OrchidPage> startIndexing() {
        Map<String, Presentation> presentations = new HashMap<>();

        if (!EdenUtils.isEmpty(sectionNames)) {
            for (String section : sectionNames) {
                Presentation presentation = getPresentation(section);
                presentations.put(section, presentation);
            }
        }

        model.initialize(presentations);

        return null;
    }

    @Override
    public void startGeneration(List<? extends OrchidPage> pages) {

    }

    private Presentation getPresentation(String section) {
        List<Slide> slides = new ArrayList<>();

        List<OrchidResource> resourcesList = context.getLocalResourceEntries(OrchidUtils.normalizePath(presentationsBaseDir) + "/" + section, null, false);
        for (OrchidResource entry : resourcesList) {
            slides.add(new Slide(entry));
        }

        return new Presentation(section, slides);
    }
}

File: plugins/OrchidPresentations/src/main/kotlin/com/eden/orchid/presentations/PresentationsGenerator.kt
package com.eden.orchid.presentations

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.generators.OrchidGenerator
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.options.annotations.StringDefault
import com.eden.orchid.api.resources.resource.OrchidResource
import com.eden.orchid.api.resources.resourcesource.LocalResourceSource
import com.eden.orchid.presentations.model.Presentation
import com.eden.orchid.presentations.model.PresentationsModel
import com.eden.orchid.presentations.model.Slide
import javax.inject.Inject

@Description("Embed presentations and slide-decks in your pages using Deck.js.", name = "Presentations")
class PresentationsGenerator : OrchidGenerator<PresentationsModel>(GENERATOR_KEY, Stage.CONTENT) {

    companion object {
        const val GENERATOR_KEY = "presentations"
    }

    @Option
    @StringDefault("presentations")
    @Description("The base directory in local resources to look for presentation slides in.")
    lateinit var baseDir: String

    override fun startIndexing(context: OrchidContext): PresentationsModel {
        return PresentationsModel(
            getPresentationResources(context, baseDir)
                .mapValues { getPresentation(context, key, it.value) }
        )
    }

    private fun getPresentationResources(context: OrchidContext, baseDir: String): Map<String, List<OrchidResource>> {
        val slides = context.getResourceEntries(baseDir, context.compilerExtensions.toTypedArray(), true, LocalResourceSource)
        val slideMap = HashMap<String, ArrayList<OrchidResource>>()

        slides.map {
            val key = it.reference.getPathSegment(1)

            it.reference.stripFromPath(baseDir)
            it.reference.stripFromPath(key)

            if (!slideMap.containsKey(key)) {
                slideMap.put(key, ArrayList())
            }
            slideMap.get(key)!!.add(it)
        }

        return slideMap
    }

    private fun getPresentation(context: OrchidContext, section: String, resources: List<OrchidResource>): Presentation {
        val slides = resources
            .mapIndexed { index, orchidResource -> Slide(context, orchidResource, index) }
            .sortedBy { it.order }

        return Presentation(section, slides, "")
    }
}

